File: dijkstra.java

import java.util.*;

public class dijkstra 
{
    public static ArrayList<ArrayList<ArrayList<Integer>>> constructAdj(int[][] edges, int V)
    {
        ArrayList<ArrayList<ArrayList<Integer>>> adj = new ArrayList<>();

        for (int i = 0; i < V; i++)
        {
            adj.add(new ArrayList<>());
        }

        for (int[] edge : edges)
        {
            int u = edge[0];
            int v = edge[1];
            int wt = edge[2];

            ArrayList<Integer> a1 = new ArrayList<>();
            a1.add(v);
            a1.add(wt);
            adj.get(u).add(a1);

            ArrayList<Integer> a2 = new ArrayList<>();
            a2.add(u);
            a2.add(wt);
            adj.get(v).add(a2);
        }
        return adj;
    }

    public static int[] dijkstraAlgo(int[][]edges, int V, int src)
    {
        ArrayList<ArrayList<ArrayList<Integer>>> adj = constructAdj(edges, V);

        PriorityQueue<ArrayList<Integer>> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.get(0)));

        int[] dist = new int[V];      
        for (int i = 0; i < V; i++)
        {
            dist[i] = Integer.MAX_VALUE;
        }

        dist[src] = 0;

        ArrayList<Integer> start = new ArrayList<>();
        start.add(src);
        start.add(0);

        pq.offer(start);

        while (!pq.isEmpty())
        {
            ArrayList<Integer> curr = pq.poll();
            int u = curr.get(1);

            for (ArrayList<Integer> neighbour : adj.get(u))
            {
                int v = neighbour.get(0);
                int wt = neighbour.get(1);

                if (dist[v] > dist[u] + wt)
                {
                    dist[v] = dist[u] + wt;

                    ArrayList<Integer> temp = new ArrayList<>();
                    temp.add(dist[v]);
                    temp.add(v);
                    pq.offer(temp);             
                } 
            }
          }
        return dist;
    }
    public static void main(String[] args) {
        int[][] edges = {{0, 1, 4}, {0, 2, 8}, {1, 4, 6},{2, 3, 2}, {3, 4, 10}};
        int V = 5;
        int src = 0;

        int[] distance = dijkstraAlgo(edges, V, src);

        for (int d : distance)
        {
            System.out.print(d + " ");
        }
    }
}

File: dijkstra.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<vector<int>>> constructAdj(vector<vector<int>> &edges, int V)
{
    vector<vector<vector<int>>> adj(V);
    for (const auto &edge : edges)
    {
        int u = edge[0];
        int v = edge[1];
        int weight = edge[2];

        adj[u].push_back({v,weight});
        adj[v].push_back({u,weight});
    }
    return adj;
}


vector<int> dijsktraAlgo(int V, vector<vector<int>> &edges, int src)
{
    vector<vector<vector<int>>> adj = constructAdj(edges, V);
    priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;
    vector<int> result(V, INT_MAX);
    pq.push({0,src});
    result[src] = 0;
    while (!pq.empty())
    {
        int u = pq.top()[1];
        pq.pop();
        for (auto neighbour : adj[u])
        {
            int v = neighbour[0];
            int weight = neighbour[1];
            if (result[v] > result[u] + weight)
            {
                result[v] = result[u] + weight;
                pq.push({result[v],v});
            }
        }
    }
    return result;
}

void printArray(vector<int> result)
{
    for (int i : result)
    {
        cout << i << " ";
    }
}

int main() {
    int V = 5;
    int src = 0;
    vector<vector<int>> edges = {{0,1,4}, {0,2,8},{1,4,6},{2,3,2}, {3,4,10}};
    vector<int> result = dijsktraAlgo(V, edges, src);
    printArray(result);
    return 0;
}
